File: lib/faye/util/web_socket/draft75_parser.rb
module Faye
  class WebSocket
    
    class Draft75Parser
      def initialize(web_socket)
        @socket    = web_socket
        @buffer    = []
        @buffering = false
      end
      
      def version
        'draft-75'
      end
      
      def handshake_response
        upgrade =  "HTTP/1.1 101 Web Socket Protocol Handshake\r\n"
        upgrade << "Upgrade: WebSocket\r\n"
        upgrade << "Connection: Upgrade\r\n"
        upgrade << "WebSocket-Origin: #{@socket.env['HTTP_ORIGIN']}\r\n"
        upgrade << "WebSocket-Location: #{@socket.url}\r\n"
        upgrade << "\r\n"
        upgrade
      end
      
      def parse(data)
        data.each_byte(&method(:handle_byte))
      end
      
      def frame(data, type = nil, error_type = nil)
        ["\x00", data, "\xFF"].map(&Faye.method(:encode)) * ''
      end
      
    private
      
      def handle_byte(data)
        case data
          when 0x00 then
            @buffering = true
            
          when 0xFF then
            @socket.receive(Faye.encode(@buffer))
            @buffer = []
            @buffering = false
            
          else
            @buffer.push(data) if @buffering
        end
      end
    end
    
  end
end

File: javascript/util/node/web_socket/draft75_parser.js
Faye.WebSocket.Draft75Parser = Faye.Class({
  FRAME_START : new Buffer([0x00]),
  FRAME_END   : new Buffer([0xFF]),
  
  version     : 'draft-75',
  
  initialize: function(webSocket, stream) {
    this._socket    = webSocket;
    this._stream    = stream;
    this._buffer    = [];
    this._buffering = false;
  },
  
  handshakeResponse: function() {
    var stream = this._stream;
    try {
      stream.write('HTTP/1.1 101 Web Socket Protocol Handshake\r\n');
      stream.write('Upgrade: WebSocket\r\n');
      stream.write('Connection: Upgrade\r\n');
      stream.write('WebSocket-Origin: ' + this._socket.request.headers.origin + '\r\n');
      stream.write('WebSocket-Location: ' + this._socket.url + '\r\n\r\n');
    } catch (e) {
      // socket closed while writing
      // no handshake sent; client will stop using WebSocket
    }
  },
  
  parse: function(data) {
    for (var i = 0, n = data.length; i < n; i++)
      this._handleChar(data[i]);
  },
  
  frame: function(data) {
    var stream = this._stream;
    try {
      stream.write(this.FRAME_START, 'binary');
      stream.write(new Buffer(data), 'utf8');
      stream.write(this.FRAME_END, 'binary');
      return true;
    } catch (e) {
      return false;
    }
  },
  
  _handleChar: function(data) {
    switch (data) {
      case 0x00:
        this._buffering = true;
        break;
      
      case 0xFF:
        this._buffer = new Buffer(this._buffer);
        this._socket.receive(this._buffer.toString('utf8', 0, this._buffer.length));
        this._buffer = [];
        this._buffering = false;
        break;
      
      default:
        if (this._buffering) this._buffer.push(data);
    }
  }
});
